### PR TITLE
S3267: Add FP repro for #7730 and FN repro for #7776

### PR DESCRIPTION
**Pull request by antonioaversa**
Add repro for https://github.com/SonarSource/sonar-dotnet/issues/7730
Add repro for https://github.com/SonarSource/sonar-dotnet/issues/7776 (extracted from 7730)

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LoopsAndLinq.CSharp11.cs [hunk 1]
```diff
@@ -1,37 +1,99 @@
...
+    }
+
+    class IterableNotEnumerable
+    {
+        public IEnumerator<char> GetEnumerator()
+        {
+            yield return 'a';
+            yield return 'b';
         }
     }
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LoopsAndLinq.CSharp11.cs:98]: Can you add also a test iterating a custom class that is inheriting a list:
```cs
var myList = new MyList();
foreach(var x in myList) { }
public class MyList : List<int>{}
```

and does this rule raises with methods/delegates?
```cs
public void ThroughLambda(Func<ImmutableList<int>> lambda)
{
    foreach(var x in MyMethod()) { }
    foreach(var x in lambda()) { }

    List<int> MyMethod() => new();
}
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LoopsAndLinq.CSharp11.cs [hunk 2]
```diff
@@ -1,37 +1,99 @@
...
+            {
+                Console.WriteLine("Pattern match successful");
             }
+
+        }
+    }
+}
+
+// https://github.com/SonarSource/sonar-dotnet/issues/7730
+class Repro_7730
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LoopsAndLinq.CSharp11.cs:37]: IMO we should split this into two different issues/reproducers:
1) 7730: FP - object being iterated should be IEnumerable/IQueryable
2) ????: FN - logical pattern matching/ logical operators

Why?
They require 2 different fixes that are not related to each other.

**Review by cristian-ambrosini-sonarsource — CHANGES_REQUESTED** [review]
I left some thoughts about splitting the issues, let me know what you think

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LoopsAndLinq.CSharp11.cs (hunk 1 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LoopsAndLinq.CSharp11.cs:98]: Added tests iterating on custom lists and also custom `IEnumerable<T>`.

Yes, the rule behaves correctly in the presence of methods, delegates, and local functions.
I have added test cases showing that.

**Review by cristian-ambrosini-sonarsource — APPROVED** [review]
LGTM! Just remember to add the link to #7776

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LoopsAndLinq.cs [hunk 3]
```diff
@@ -280,7 +280,35 @@ public void ForEach_Compliant(ICollection<string> collection, ICollection<Point>
             }
         }
 
-        public void M(DataTable rawSheetData)
+        public void ForEach_WithListAndLogicalOperators(List<char> s)
+        {
+            foreach (var c in s)                   // FN, equivalent to c is ' '
```
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LoopsAndLinq.cs]: Done, extracted into a dedicated `Repro_7766` class.
